handler.go

```diff
@@ -0,0 +1,214 @@
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// BackendReader is the interface that wraps the Read method.
+//
+// Backends implement this interface to provide DHCP data to the DHCP server.
+type BackendReader interface {
+	// Read data (from a backend) based on a mac address
+	// and return DHCP headers and options, including netboot info.
+	Read(context.Context, net.HardwareAddr) (*data.Dhcp, *data.Netboot, error)
+}
+
+// handleFunc is the main handler for DHCPv4 packets.
+func (s *Server) handleFunc(conn net.PacketConn, peer net.Addr, m *dhcpv4.DHCPv4) {
+	log := s.Log.WithValues("mac", m.ClientHWAddr.String())
+	tracer := otel.Tracer("DHCP")
+	ctx, span := tracer.Start(s.ctx, "DHCP Packet Received",
+		trace.WithAttributes(attribute.String("MAC", m.ClientHWAddr.String())),
+		trace.WithAttributes(attribute.String("MessageType", m.MessageType().String())),
+	)
+	defer span.End()
+	var reply *dhcpv4.DHCPv4
+	switch mt := m.MessageType(); mt {
+	case dhcpv4.MessageTypeDiscover:
+		reply = s.handleDiscover(ctx, tracer, m)
+	case dhcpv4.MessageTypeRequest:
+		reply = s.handleRequest(ctx, tracer, m)
+	case dhcpv4.MessageTypeRelease:
+		s.handleRelease(ctx, m)
+	default:
+		log.Info("received unknown message type", "type", mt)
+	}
+	c := codes.Ok
+	cmsg := fmt.Sprintf("no reply send DHCP: %v", m.MessageType())
+	if reply != nil {
+		span.SetAttributes(attribute.String("IP", reply.YourIPAddr.String()))
+		if _, err := conn.WriteTo(reply.ToBytes(), peer); err != nil {
+			log.Error(err, "failed to send DHCP")
+			c = codes.Error
+			cmsg = fmt.Sprintf("failed to send DHCP: %v, err: %v", reply.MessageType().String(), err.Error())
+		} else {
+			log.Info("sent DHCP response")
+			cmsg = fmt.Sprintf("sent DHCP: %s", reply.MessageType().String())
+		}
+	}
+	span.SetStatus(c, cmsg)
+}
+
+// handleDiscover handles DHCP packets message type of discover.
+func (s *Server) handleDiscover(ctx context.Context, tracer trace.Tracer, m *dhcpv4.DHCPv4) *dhcpv4.DHCPv4 {
+	log := s.Log.WithValues("mac", m.ClientHWAddr.String())
+	ctx, span := tracer.Start(ctx, "DHCP Discover Message")
+	log.Info("received discover packet")
+	d, n, err := s.Backend.Read(ctx, m.ClientHWAddr)
+	if err != nil {
+		s.Log.Info("not sending DHCP OFFER", "mac", m.ClientHWAddr, "error", err)
+		span.SetStatus(codes.Error, fmt.Sprintf("not sending DHCP OFFER: %v", err))
+		span.End()
+		return nil
+	}
+	span.SetAttributes(
+		attribute.String("Subnet", d.SubnetMask.String()),
+		attribute.String("DefaultGateway", d.DefaultGateway.String()),
+		attribute.String("Nameservers", fmt.Sprintf("%v", d.NameServers)),
+		attribute.String("Hostname", d.Hostname),
+		attribute.String("DomainName", d.DomainName),
+		attribute.Int("LeaseTime", int(d.LeaseTime)),
+		attribute.Bool("AllowNetboot", n.AllowNetboot),
+		attribute.String("IpxeScriptURL", fmt.Sprintf("%v", n.IpxeScriptURL)),
+	)
+	span.SetStatus(codes.Ok, "done reading from backend")
+	span.End()
```

Comment:
Perhaps `span.End()` should be called with `defer`?
Also, the "done reading from backend" doesn't quite seem to match "DHCP Discover Message".
I think the tracing in this function could be a little better.

Reply:
yep, in Go, you typically want to defer span.End unless there's some reason you'd want to drop the span (bc span.End is when they actually get sent)